libs/ledger/src/protocols/main_chain_rpc_service.cpp

```diff
@@ -240,11 +240,18 @@ MainChainRpcService::Address MainChainRpcService::GetRandomTrustedPeer() const
 
   if (!direct_peers.empty())
   {
-    // generate a random peer index
-    std::size_t const index = rng() % direct_peers.size();
+    if (alien_peers_.size() == direct_peers.size())
+    {
+      return {};
+    }
+    do
+    {
+      // generate a random peer index
+      std::size_t const index = rng() % direct_peers.size();
 
-    // select the address
-    address = direct_peers[index];
+      // select the address
+      address = direct_peers[index];
+    } while (alien_peers_.IsBlacklisted(address));
```

Comment:
use `std::find_if_not`

Reply:
> use std::find_if_not

Hmm, @wkaluza, do you suggest employing a custom randomized `ExecutionPolicy`? One problem is, std algorithms accepting policies were only standardised in 17 (and even then most STL implementations did not support them). Even then, that would be an overcomplication of random choice, IMO.

Reply:
You're right, the randomisation would be tricky. Let's leave it as it is, then.